hathor/indexes/rocksdb_timestamp_index.py

```diff
@@ -0,0 +1,153 @@
+              *, reverse: bool = False) -> Iterator[Tuple[int, bytes]]:
+        """ Iterate over transactions optionally starting from a timestamp/hash, by default from oldest to newest.
+        """
+        if from_timestamp is None and from_tx is not None:
+            raise ValueError('from_tx needs from_timestamp, but it is None')
+        it = self._db.iterkeys(self._cf)
+        if reverse:
+            it = reversed(it)
+            if from_timestamp is None:
+                self.log.debug('seek to last')
+                it.seek_to_last()
+            else:
+                # when reversed we increment the key by 1, which effectively goes to the end of a prefix
+                self.log.debug('seek to', tx=from_tx)
+                it.seek_for_prev(incr_key(self._to_key(from_timestamp, from_tx)))
+        else:
+            if from_timestamp is None:
+                self.log.debug('seek to first')
+                it.seek_to_first()
+            else:
+                self.log.debug('seek to', tx=from_tx)
+                it.seek(self._to_key(from_timestamp, from_tx))
+        first = True
+        for _, key in it:
+            timestamp, tx_hash = self._from_key(key)
+            self.log.debug('seek found', tx=tx_hash.hex())
+            # XXX: this is made such that we wouldn't skip the first element blindly, only if it is the given from_tx
+            if first:
+                first = False
+                # from_tx transaction is never yielded
+                if from_timestamp is not None and from_timestamp == timestamp and from_tx == tx_hash:
+                    continue
+            yield timestamp, tx_hash
+        self.log.debug('seek end')
+
+    def get_newest(self, count: int) -> Tuple[List[bytes], bool]:
+        it = (x for _, x in self._iter(reverse=True))
+        return collect_n(it, count)
+
+    def get_older(self, timestamp: int, hash_bytes: bytes, count: int) -> Tuple[List[bytes], bool]:
+        it = (x for _, x in self._iter(timestamp, hash_bytes, reverse=True))
+        return collect_n(it, count)
+
+    def get_newer(self, timestamp: int, hash_bytes: bytes, count: int) -> Tuple[List[bytes], bool]:
+        it = (x for _, x in self._iter(timestamp, hash_bytes))
+        return collect_n(it, count)
+
+    def get_hashes_and_next_idx(self, from_idx: RangeIdx, count: int) -> Tuple[List[bytes], Optional[RangeIdx]]:
+        assert count > 0, 'count<=0 not allowed'
+        timestamp, offset = from_idx
+        it = skip_n(self._iter(timestamp), offset)
+        hashes: List[bytes] = []
+        n = count
+        ret_next_timestamp = None
+        ret_next_offset = offset
+        while n > 0:
+            try:
+                next_timestamp, tx_hash = next(it)
+            except StopIteration:
+                return hashes, None
+            hashes.append(tx_hash)
+            if ret_next_timestamp != next_timestamp:
+                if ret_next_timestamp is not None:
```

Comment:
I think this still has a bug but I'm not sure. Imagine if we go through 3 different timestamps, you would be setting `ret_next_offset = 0` only in the first change because after that `ret_next_timestamp` would never be `None` again.

I think you can fix that with the following code

```python
ret_next_timestamp = timestamp
ret_next_offset = offset
while n > 0:
...
    if ret_next_timestamp != next_timestamp:
        ret_next_offset = 0
        ret_next_timestamp = next_timestamp
    else:
    ....
```

What do you think?

Reply:
Oh, got it. You're right. The use of `None` isn't really adding anything. I'll go with your suggestion.

Reply:
__Now__ it's fixed, and the test actually fails on any of the wrong variations we discussed before.